google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java

```diff
@@ -550,15 +638,59 @@ public void individualWritesErrorIfBulkCommitFails() throws Exception {
     ApiFuture<WriteResult> result2 = bulkWriter.set(doc2, LocalFirestoreHelper.SINGLE_FIELD_MAP);
     bulkWriter.close().get();
 
-    String expected = "java.lang.IllegalStateException: Mock batchWrite failed in test";
     for (ApiFuture<WriteResult> result : Arrays.asList(result1, result2)) {
       try {
         result.get();
       } catch (Exception e) {
-        assertEquals(expected, e.getMessage());
+        assertTrue(
+            e.getMessage()
+                .contains("java.lang.IllegalStateException: Mock batchWrite failed in test"));
         ++opCount;
       }
     }
     assertEquals(2, opCount);
   }
+
+  @Test
+  public void retriesWritesWhenBatchWriteFailsWithRetryableError() throws Exception {
+    FirestoreException retryableError =
+        FirestoreException.serverRejected(
+            Status.fromCode(Status.Code.ABORTED), "Mock batchWrite failed in test");
+
+    doThrow(retryableError)
+        .doThrow(retryableError)
+        .doThrow(retryableError)
+        .doReturn(successResponse(3))
+        .when(firestoreMock)
+        .sendRequest(
+            batchWriteCapture.capture(),
+            Matchers.<UnaryCallable<BatchWriteRequest, BatchWriteResponse>>any());
+
+    ApiFuture<WriteResult> result = bulkWriter.set(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+    bulkWriter.close().get();
+
+    assertEquals(Timestamp.ofTimeSecondsAndNanos(3, 0), result.get().getUpdateTime());
+  }
+
+  @Test
+  public void failsWritesAfterAllRetryAttemptsFail() throws Exception {
+    doThrow(
```

Comment:
Does the mocking layer allow us to specify that we want this function to be called 10 times?

Reply:
Changed to use doAnswer.